File: backend/app/Services/TripService.php
<?php

namespace App\Services;

use App\Enums\BookingStatus;
use App\Enums\TripStatus;
use App\Models\Booking;
use App\Models\Inspection;
use App\Models\Trip;

class TripService
{
    public function startTrip(Booking $booking, array $inspectionData = []): Trip
    {
        if ($booking->status !== BookingStatus::Confirmed) {
            throw new \RuntimeException('Booking must be confirmed to start a trip.');
        }

        $trip = Trip::create([
            'booking_id' => $booking->id,
            'status' => TripStatus::InProgress,
            'mileage_start' => $inspectionData['mileage'] ?? null,
            'fuel_level_start' => $inspectionData['fuel_level'] ?? null,
            'started_at' => now(),
        ]);

        $booking->update(['status' => BookingStatus::Active]);

        if (!empty($inspectionData)) {
            $this->createInspection($trip, 'pre_trip', $inspectionData);
        }

        return $trip;
    }

    public function endTrip(Trip $trip, array $inspectionData = []): Trip
    {
        if ($trip->status !== TripStatus::InProgress) {
            throw new \RuntimeException('Trip is not currently in progress.');
        }

        $trip->update([
            'status' => TripStatus::Completed,
            'mileage_end' => $inspectionData['mileage'] ?? null,
            'fuel_level_end' => $inspectionData['fuel_level'] ?? null,
            'ended_at' => now(),
        ]);

        if (!empty($inspectionData)) {
            $this->createInspection($trip, 'post_trip', $inspectionData);
        }

        return $trip->fresh();
    }

    public function createInspection(Trip $trip, string $type, array $data): Inspection
    {
        return Inspection::create([
            'trip_id' => $trip->id,
            'type' => $type,
            'zones' => $data['zones'] ?? [],
            'photo_urls' => $data['photo_urls'] ?? [],
            'notes' => $data['notes'] ?? null,
            'mileage' => $data['mileage'] ?? null,
            'fuel_level' => $data['fuel_level'] ?? null,
            'damage_reported' => $data['damage_reported'] ?? false,
            'damage_description' => $data['damage_description'] ?? null,
        ]);
    }

    public function updateLocation(Trip $trip, float $lat, float $lng): Trip
    {
        $trip->update([
            'last_known_lat' => $lat,
            'last_known_lng' => $lng,
            'location_updated_at' => now(),
        ]);

        return $trip;
    }

    public function extendTrip(Trip $trip, int $extraDays): Trip
    {
        $trip->update([
            'extended' => true,
            'extension_days' => ($trip->extension_days ?? 0) + $extraDays,
        ]);

        $booking = $trip->booking;
        $newReturnAt = $booking->return_at->addDays($extraDays);
        $booking->update(['return_at' => $newReturnAt]);

        return $trip->fresh();
    }
}

File: frontend/lib/features/trip/data/trip_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../../core/constants/api_endpoints.dart';
import '../../../core/models/trip.dart';
import '../../../core/network/api_response.dart';
import '../../../core/network/dio_client.dart';
import '../../../core/network/error_handler.dart';

final tripServiceProvider = Provider<TripService>((ref) {
  return TripService(ref.read(dioProvider));
});

/// Trip lifecycle endpoints (start/end/extend/location).
class TripService {
  final Dio _dio;
  TripService(this._dio);

  Future<Trip> get(String tripId) async {
    try {
      final res = await _dio.get(ApiEndpoints.tripDetail(tripId));
      return Trip.fromJson(ApiResponse.data(res.data));
    } catch (e) {
      throw mapError(e);
    }
  }

  Future<Trip> start(
    String bookingId, {
    int? mileage,
    int? fuelLevel,
    List<String>? photoUrls,
    String? notes,
  }) async {
    try {
      final res = await _dio.post(ApiEndpoints.startTrip(bookingId), data: {
        if (mileage != null) 'mileage': mileage,
        if (fuelLevel != null) 'fuel_level': fuelLevel,
        if (photoUrls != null) 'photo_urls': photoUrls,
        if (notes != null) 'notes': notes,
      });
      return Trip.fromJson(ApiResponse.data(res.data));
    } catch (e) {
      throw mapError(e);
    }
  }

  Future<Trip> end(
    String tripId, {
    int? mileage,
    int? fuelLevel,
    List<String>? photoUrls,
    String? notes,
    bool? damageReported,
    String? damageDescription,
  }) async {
    try {
      final res = await _dio.post(ApiEndpoints.endTrip(tripId), data: {
        if (mileage != null) 'mileage': mileage,
        if (fuelLevel != null) 'fuel_level': fuelLevel,
        if (photoUrls != null) 'photo_urls': photoUrls,
        if (notes != null) 'notes': notes,
        if (damageReported != null) 'damage_reported': damageReported,
        if (damageDescription != null) 'damage_description': damageDescription,
      });
      return Trip.fromJson(ApiResponse.data(res.data));
    } catch (e) {
      throw mapError(e);
    }
  }

  Future<void> updateLocation(String tripId, double lat, double lng) async {
    try {
      await _dio.post(
        ApiEndpoints.updateTripLocation(tripId),
        data: {'lat': lat, 'lng': lng},
      );
    } catch (e) {
      throw mapError(e);
    }
  }

  Future<Trip> extend(String tripId, int extraDays) async {
    try {
      final res = await _dio.post(
        ApiEndpoints.extendTrip(tripId),
        data: {'extra_days': extraDays},
      );
      return Trip.fromJson(ApiResponse.data(res.data));
    } catch (e) {
      throw mapError(e);
    }
  }
}
